### PR TITLE
[multistage] [testing] Add join operator unit test

### PR DESCRIPTION
**Pull request by 61yao**

### CODE REVIEW

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java [hunk 1]
```diff
@@ -26,6 +26,11 @@ public class QueryTestSet {
   @DataProvider(name = "testSql")
   public Object[][] provideTestSql() {
     return new Object[][]{
+        // Failed left join. Not sure what's the expectation here.
+        // new Object[]{"SELECT a.col1 FROM a LEFT JOIN b ON (a.col1 = 'foo')"},
+        // new Object[]{"SELECT a.col1 FROM a LEFT JOIN b ON (a.col1 = b.col1 and a.col1 = 'foo)"},
+        // Failed semi join. Somehow it gets time out now
+        // new Object[]{"SELECT * FROM b WHERE b.col1 IN (SELECT a.col1 FROM a)"},
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java:33]: let's dont add these to the E2E test. but to the operator test

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java [hunk 2]
```diff
@@ -42,80 +69,323 @@ private static JoinNode.JoinKeys getJoinKeys(List<Integer> leftIdx, List<Integer
...
     while (result.isNoOpBlock()) {
       result = join.nextBlock();
     }
     List<Object[]> resultRows = result.getContainer();
-    Object[] expRow = new Object[]{1, "Aa", 2, "Aa"};
-    List<Object[]> expectedRows = new ArrayList<>();
-    expectedRows.add(expRow);
-    Assert.assertEquals(expectedRows.size(), resultRows.size());
-    Assert.assertEquals(expectedRows.get(0), resultRows.get(0));
+    List<Object[]> expectedRows = Arrays.asList(new Object[]{1, "Aa", 2, "Aa"}, new Object[]{2, "CC", null, null});
```
Comment (agavra) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java:207]: can we also add a test for 1:N left join? (e.g. one record on the left joins with multiple on the right)

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java [hunk 3]
```diff
@@ -22,18 +22,45 @@
...
+import org.mockito.MockitoAnnotations;
 import org.testng.Assert;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.Test;
 
 
-
 public class HashJoinOperatorTest {
+  private AutoCloseable _mocks;
```
Comment (agavra) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java:45]: can we add tests for handling upstream NOOP, EOS (with no other rows, e.g. one side is empty) and ERROR?

**Review by agavra — APPROVED** [review]
Thanks @61yao! LGTM with some comments for extra coverage

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java [hunk 4]
```diff
@@ -26,6 +26,11 @@ public class QueryTestSet {
   @DataProvider(name = "testSql")
   public Object[][] provideTestSql() {
     return new Object[][]{
+        // Failed left join. Not sure what's the expectation here.
+        // new Object[]{"SELECT a.col1 FROM a LEFT JOIN b ON (a.col1 = 'foo')"},
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java:30]: also FYI this is a cross join (since there's no rule for b)

Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java:30]: Removed.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java (hunk 3 above)
Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java:45]: Yeah. Sorry, somehow I forgot to add the corner case testing. It just slipped. Added the coverage and found a bug!

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java (hunk 2 above)
Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/HashJoinOperatorTest.java:207]: Done

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java (hunk 1 above)
Comment (61yao) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/QueryTestSet.java:33]: Removed

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 5]
```diff
@@ -63,13 +64,18 @@ public class HashJoinOperator extends BaseOperator<TransferableBlock> {
...
-  // TODO: Double check semi join logic.
   public HashJoinOperator(Operator<TransferableBlock> leftTableOperator, Operator<TransferableBlock> rightTableOperator,
       DataSchema outputSchema, JoinNode.JoinKeys joinKeys, List<RexExpression> joinClauses, JoinRelType joinType) {
-    // TODO: Handle the case where _leftKeySelector and _rightKeySelector could be null.
+    // TODO: Move precondition check to constructor and handle error in upstream.
+    // TODO: Fix semi join bug. (https://github.com/apache/pinot/issues/9757)
+    Preconditions.checkState(joinType != JoinRelType.SEMI, "Semi join is not supported");
+    Preconditions.checkState(joinType != JoinRelType.RIGHT, "Right join is not supported");
+    Preconditions.checkState(joinType != JoinRelType.FULL, "Full join is not supported");
+    Preconditions.checkState(joinType != JoinRelType.ANTI, "Anti join is not supported");
```
[the remark this answers is not in the document]
Comment (61yao) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:74]: Done

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 6]
```diff
@@ -82,6 +90,8 @@ public HashJoinOperator(Operator<TransferableBlock> leftTableOperator, Operator<
     _isHashTableBuilt = false;
     _broadcastHashTable = new HashMap<>();
     _upstreamErrorBlock = null;
+    // TODO: Consider figuring out a better way to support join data left empty case.
+    _hasSentData = false;
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:94]: i dont think this is actually a bug on the left empty or right empty. processing of a left or right empty block is the same as not-processing any block. and end-of-stream is a pass through. 

I agree we should consolidate the behavior but let's not address this issue here and split into a separate PR

**Review by walterddr — COMMENTED** [review]
lgtm overall. one last comment on the empty issue.

**Review by walterddr — APPROVED** [review]
